machetli/pddl/files.py

```diff
@@ -55,6 +82,56 @@ def temporary_files(state: dict) -> tuple:
     os.remove(problem_file.name)
 
 
+def run_evaluator(evaluate):
+    """
+    Load the state passed to the script via its command line arguments, then run
+    the given function *evaluate* on the domain and problem encoded in the
+    state, and exit the program with the appropriate exit code. If the function
+    returns ``True``, use
+    :attr:`EXIT_CODE_IMPROVING<machetli.evaluator.EXIT_CODE_IMPROVING>`
+    otherwise, use
+    :attr:`EXIT_CODE_NOT_IMPROVING<machetli.evaluator.EXIT_CODE_NOT_IMPROVING>`.
+
+    This function is meant to be used as the main function of an evaluator
+    script. Instead of a path to the state, the command line arguments can also
+    be paths to a PDDL domain and problem (where the domain can be omitted if it
+    can be found with automated naming rules). This is meant for testing and
+    debugging the evaluator directly on PDDL input.
+
+    :param evaluate: is a function taking filenames of a PDDL domain and problem
+        file as input and returning ``True`` if the specified behavior occurs for
+        the given instance, and ``False`` if it doesn't. Other ways of exiting the
+        function (exceptions, ``sys.exit`` with exit codes other than
+        :attr:`EXIT_CODE_IMPROVING<machetli.evaluator.EXIT_CODE_IMPROVING>` or
+        :attr:`EXIT_CODE_NOT_IMPROVING<machetli.evaluator.EXIT_CODE_NOT_IMPROVING>`)
+        are treated as failed evaluations by the search.
+    """
+    filenames = sys.argv[1:]
+    if len(filenames) == 1:
+        try:
+            state = tools.read_state(filenames[0])
+        except (FileNotFoundError, PickleError):
+            task_filename = filenames[0]
+            domain_filename = _find_domain_filename(task_filename)
+            state = generate_initial_state(domain_filename, task_filename)
+    elif len(filenames) == 2:
+        domain_filename, task_filename = filenames
+        state = generate_initial_state(domain_filename, task_filename)
+    else:
+        logging.critical(
+            "Error: evaluator has to be called with either a path to a pickled state, "
+            "a task filename, or a domain filename followed by a task filename.")
+        sys.exit(EXIT_CODE_CRITICAL)
+
+    with temporary_files(state) as (domain_filename, task_filename):
+        improving = evaluate(domain_filename, task_filename)
```

Comment:
I thought this way would be cleaner because we treat everything the same once we have the state. But you're right, it's not necessary. I want to avoid too much logic inside the `except` clause. or duplicating the code for reacting to the value of `evaluate`. Can you make a suggestion for a rewrite?